metadata/src/utils/retain.rs

```diff
@@ -8,45 +8,185 @@ use crate::{
     ExtrinsicMetadata, Metadata, OuterEnumsMetadata, PalletMetadataInner, RuntimeApiMetadataInner,
     StorageEntryType,
 };
-use alloc::collections::BTreeMap;
-use hashbrown::HashSet;
-use scale_info::TypeDef;
-
-/// Collect all type IDs needed to represent the provided pallet.
-fn collect_pallet_types(pallet: &PalletMetadataInner, type_ids: &mut HashSet<u32>) {
-    if let Some(storage) = &pallet.storage {
-        for entry in storage.entries() {
-            match entry.entry_type {
-                StorageEntryType::Plain(ty) => {
-                    type_ids.insert(ty);
+use alloc::collections::{BTreeMap, BTreeSet, VecDeque};
+use alloc::vec::Vec;
+use scale_info::{
+    PortableType, TypeDef, TypeDefArray, TypeDefBitSequence, TypeDefCompact, TypeDefComposite,
+    TypeDefSequence, TypeDefTuple, TypeDefVariant,
+};
+
+#[derive(Clone)]
+struct TypeSet {
```

Comment:
Personally I'd likely make the functions impl'd on this free standing and just have a. `BTreeSet` passed about,  primarily for the reason that despite this `TypeSet` struct, there are still places where you manually do `type_set.seen_ids.insert(foo)` or whatever, so the internal details are leaking anyway :)

Reply:
removed those parts to use defined method on the struct. i moved those functions an impl block to not carry the set around :D